Fix hopper lag by making entities look for hoppers.

Techcable opened the pull request:
Every tick hoppers try and find an block-inventory to extract from.
If no tile entity is above the hopper (which there often isn't) it will do a bounding box search for minecart chests and minecart hoppers.
If it can't find an inventory, it will then look for a dropped item, which is another bounding box search.
This patch eliminates that expensive check by having dropped items and minecart hoppers/chests look for hoppers instead.
Hoppers are tile entities meaning you can do a simple tile entity lookup to find the nearest hopper in range.
This usually results in a net performance gain, since tile entity lookups are around a order of magnitude faster then bounding box lookups.
Pushing out of hoppers also causes a unessicary bouding box lookup, which this patch replaces with a tile entity lookup.

This patch may causes a decrease in the performance of dropped items, which is why it can be disabled in the configuration.
# 

This patch has gone through signifigant improvements since you [first saw it a year ago](https://hub.spigotmc.org/stash/projects/PAPER/repos/paperspigot/pull-requests/9/overview).
- It is now fully compatible with `useHopperCheck` and preserves vanilla lookup behavior for items and minecarts.
- It has survived a major mincraft update with no major changes, and multple minor updates.
  - The only issue was that it conflicted with @Aikar 's hopper cooldown reset, but that was later fixed in TacoSpigot#TacoSpigot@577f2ec93a57c7767acbf2605ba33ffd2928ecf9
- Should emulate vanilla behavior perfectly
- Used in production on [over 50 servers serving over 100 players](http://mcstats.org/plugin/TacoSpigot)
  - This includes @L4BORG, @PhanaticD and @CullanP 's servers
- It has been refactored to use java 8's default interface methods per @0x277F 's suggestion
- Many of its unnecessary diffs were removed (like that one with the Chunk.tileEntityMap using trove)

Techcable committed:
Fix hopper lag by making entities look for hoppers.

Every tick hoppers try and find an block-inventory to extract from.
If no tile entity is above the hopper (which there often isn't) it will do a bounding box search for minecart chests and minecart hoppers.
If it can't find an inventory, it will then look for a dropped item, which is another bounding box search.
This patch eliminates that expensive check by having dropped items and minecart hoppers/chests look for hoppers instead.
Hoppers are tile entities meaning you can do a simple tile entity lookup to find the nearest hopper in range.
This usually results in a net performance gain, since tile entity lookups are around a order of magnitude faster then bounding box lookups.
Pushing out of hoppers also causes a unessicary bouding box lookup, which this patch replaces with a tile entity lookup.

This patch may causes a decrease in the performance of dropped items, which is why it can be disabled in the configuration.

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++
++public interface HopperPusher {
++
++    public default TileEntityHopper findHopper() {
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++        return null;
++    }
++
++    public boolean acceptItem(TileEntityHopper hopper);
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++
++    public boolean acceptItem(TileEntityHopper hopper);
++
++    public default boolean tryPutInHopper() {
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++        return hopper != null && hopper.canAcceptItems() && acceptItem(hopper);
++    }
++
++    public AxisAlignedBB getBoundingBox();
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++
++    // Default implementations for entities
++
++    public default double getX() {
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++
++    public AxisAlignedBB getBoundingBox();
++
++    public World getWorld();
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++        return ((Entity) this).locY;
++    }
++
++    public default double getZ() {
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++        return ((Entity) this).locX;
++    }
++
++    public default double getY() {
```
Redundant `public`

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++    // Default implementations for entities
++
++    default double getX() {
++        return ((Entity) this).locX;
```
Perhaps document (in javadoc) that this interface assumes it's being applied to `Entity`.

aikar commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
+ 
+     private boolean H() {
+-        IInventory iinventory = this.I();
++        IInventory iinventory = HopperHelper.getInventory(getWorld(), getPosition().shift(BlockHopper.e(this.u()))); // Paper - avoid bounding box search
```
need obf helpers here

aikar replied:
Actually this line doesnt look good. this results in a method that does block lookups. You should just make it so that it doesnt call the getEntities method if push based is enabled on 
public static IInventory b(World world, double d0, double d1, double d2) {

And avoid all this obfuscated crap being duplicated.

aikar commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++    // Paper start
++    public AxisAlignedBB getHopperLookupBoundingBox() {
++        // Change this if b(IHopper) ever changes
++        return getHopperLookupBoundingBox(this.E(), this.F(), this.G());
```
As I did in my fork, lets add obf helpers to IHopper like so:

double E();default double getX() { return E(); } // Paper OBFHELPER

These stupid getters have caused me sooo many conflicts over the years.

aikar commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,326 @@
++
++    private static AxisAlignedBB getHopperLookupBoundingBox(double d0, double d1, double d2) {
++        // Change this if a(World, double, double, double) above ever changes
++        return new AxisAlignedBB(d0 - 0.5D, d1, d2 - 0.5D, d0 + 0.5D, d1 + 1.5D, d2 + 0.5D);
```
Add a comment on said line to force a conflict

Techcable committed:
Add more OBFHELPERs and try to tie the inventory lookup logic to NMS (triggering conflicts)

Add getX(), getY(), getZ() methods to Entity to implement HopperPusher, and remove the default implementations (those casts are yucky).
Add a flag to the IInventory lookup method (TileEntityHopper.b(IHopper)) on weather or not it should do a bounding box lookup, instead of reimplementing the logic entirely.

kashike commented on Spigot-Server-Patches/0004-MC-Utils.patch:
```diff
@@ -1,16 +1,29 @@
+ 
+     @Nullable
+-    public TileEntity a(BlockPosition blockposition, Chunk.EnumTileEntityState chunk_enumtileentitystate) {
++    public final TileEntity getTileEntity(BlockPosition pos) { return a(pos, EnumTileEntityState.IMMEDIATE); } @Nullable public TileEntity a(BlockPosition blockposition, Chunk.EnumTileEntityState chunk_enumtileentitystate) { // Paper - OBFHELPER
```
It would be nice to have it formatted like:

``` java
    @Nullable public final TileEntity getTileEntity(BlockPosition pos) { return a(pos, EnumTileEntityState.IMMEDIATE); }  // Paper - OBFHELPER
    @Nullable
    public TileEntity a(BlockPosition blockposition, Chunk.EnumTileEntityState chunk_enumtileentitystate) {
```

kashike replied:
Additionally, name it `getTileEntityImmediate` since this operates using `IMMEDIATE`.

Techcable replied:
Would this trigger a conflict when you wiggle it?

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,336 @@
++        for (int x = startX; x <= endX; x++) {
++            for (int y = startY; y <= endY; y++) {
++                for (int z = startZ; z <= endZ; z++) {
++                    BlockPosition adjacentPos = new BlockPosition(x, y, z);
```
You can probably use a `MutableBlockPos` here (create alongside `endZ`, set xyz here instead)

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,336 @@
+     public double lastX;
+     public double lastY;
+     public double lastZ;
++    // Paper start - getters to implement HopperPusher
```
This start line should be moved above `public double getX() {`.

kashike commented on Spigot-Server-Patches/0165-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch:
```diff
@@ -0,0 +1,336 @@
+     World getWorld();
+ 
+-    double E();
++    double E(); default double getX() { return E(); } // Paper - OBFHELPER
```
See other comment regarding placement of this

Techcable replied:
@aikar wanted me to do it like this

aikar replied:
for simpler lines this is good to force a conflict, but for larger lines where it hurts readability, i tend to prefer the other method kash mentioned for the method.

kashike replied:
ex: https://github.com/PaperMC/Paper/pull/319#discussion_r65823055

Techcable committed:
Use PooledBlockPosition in HopperPusher.findHopper()

Make some formatting changes to the OBFHELPERs requested by @kashike

Techcable committed:
Cleanup/Reformat the OBFHELPER for getTileEntityImmediately

Previously named getTileEntity(), which didn't properly note that EnumTileEntityState was IMMEDIATE
